src/storage/src/render/persist_sink.rs

```diff
@@ -7,113 +7,824 @@
+
+    // Only one worker is responsible for determining batch descriptions. All
+    // workers must write batches with the same description, to ensure that they
+    // can be combined into one batch that gets appended to Consensus state.
+    let hashed_id = collection_id.hashed();
+    let active_worker = usize::cast_from(hashed_id) % scope.peers() == scope.index();
+
+    // Only the "active" operator will mint batches. All other workers have an
+    // empty frontier. It's necessary to insert all of these into
+    // `compute_state.sink_write_frontier` below so we properly clear out
+    // default frontiers of non-active workers.
+
+    let mut mint_op = AsyncOperatorBuilder::new(
+        format!("{} mint_batch_descriptions", operator_name),
+        scope.clone(),
+    );
+
+    let (mut output, output_stream) = mint_op.new_output();
+
+    let mut desired_input = mint_op.new_input(desired_stream, Pipeline);
+
+    let shutdown_button = mint_op.build(move |mut capabilities| async move {
+        let mut cap_set = CapabilitySet::from_elem(capabilities.pop().expect("missing capability"));
+
+        // Initialize this operators's `upper` to the `upper` of the persist shard we are writing
+        // to. Data from the source not beyond this time will be dropped, as it has already
+        // been persisted.
+        // In the future, sources will avoid passing through data not beyond this upper
+        let mut current_upper = {
+            // TODO(aljoscha): We need to figure out what to do with error
+            // results from these calls.
+            let persist_client = persist_clients
+                .open(persist_location)
+                .await
+                .expect("could not open persist client");
+
+            let write = persist_client
+                .open_writer::<SourceData, (), Timestamp, Diff>(
+                    shard_id,
+                    &format!(
+                        "compute::persist_sink::mint_batch_descriptions {}",
+                        collection_id
+                    ),
+                    Arc::new(target_relation_desc),
+                    Arc::new(UnitSchema),
+                )
+                .await
+                .expect("could not open persist shard");
+
+            if active_worker {
+                // VERY IMPORTANT: Only the active write worker produces
+                // batch descriptions. All other worker simply drop their capability
+                // and stop working.
+                write.upper().clone()
+            } else {
+                // The non-active workers report that they are done snapshotting.
+                return;
+            }
+        };
+
+        // The current input frontiers.
+        let mut desired_frontier;
+
+        loop {
+            if let Some(event) = desired_input.next().await {
+                match event {
+                    Event::Data(_cap, _data) => {
+                        // Just read away data.
```

Comment:
This unfortunately creates a copy of the data which is then discarded immediately. An alternative would be for this operator to have two outputs where one is the data and the other is the batch description. The underlying issue is that there are multiple consumers of a stream, which means that all but one receive a copy of the data. Can explain more if needed!

Reply:
Interesting! Makes sense. So you're saying instead of splitting the input stream to both operator 1 and 2, pipeline the data through op1 into op2 so op1 can just forward the data along? I just checked and it seems like the current compute persist_sink might be doing the bad thing here, can you confirm?

Reply:
this copy shouldn't be _too_ bad in the storage case, because the input is pipelined...

A second output is a good idea, but we need to be careful about output frontiers!